LDAP Provider: Added auto-generated uidNumber and guidNumber generated attributes for use with SSSD and similar software.

Rynoxx opened the pull request:
The starting number for uid/gid can be configured iva environtment
variables and is by default 2000 which should work fine for most instances unless there are more than
999 local accounts on the server/computer.

The uidNumber is just the users Pk + the starting number.
The guidNumber is calculated by the last couple of bytes in the uuid of
the group + the starting number, this should have a low enough chance
for collisions that it's going to be fine for most use cases.

I have not added any interface stuff for configuring the environment variables as I couldn't really find my way around all the places I'd have to edit to add it and the default values should in my opinion be fine for 99% use cases.

Rynoxx committed:
Added auto-generated uidNumber and guidNumber generated attributes for
use with SSSD and similar software.

The starting number for uid/gid can be configured iva environtment
variables and is by default 2000 which should work fine for most instances unless there are more than
999 local accounts on the server/computer.

The uidNumber is just the users Pk + the starting number.
The guidNumber is calculated by the last couple of bytes in the uuid of
the group + the starting number, this should have a low enough chance
for collisions that it's going to be fine for most use cases.

I have not added any interface stuff for configuring the environment variables as I couldn't really find my way around all the places I'd have to edit to add it and the default values should in my opinion be fine for 99% use cases.

Rynoxx committed:
Add a 'fake' primary group for each user

Rynoxx committed:
Merge branch 'master' into ldap-autogenerated-idNumber

Rynoxx committed:
First attempt att adding config to interface

Rynoxx committed:
Merge conflicts from master

Rynoxx committed:
Updated API to support new fields

BeryJu commented on outpost/pkg/ldap/instance_search.go:
```diff
@@ -137,3 +159,32 @@ func (pi *ProviderInstance) GroupEntry(g api.Group) *ldap.Entry {
+		},
+		{
+			Name:   "objectClass",
+			Values: []string{GroupObjectClass, "goauthentik.io/ldap/group"},
```
Maybe we can add another objectClass to indicate that this is a fake group, like `goauthentik.io/ldap/virtual-group`?

Rynoxx replied:
Instead of `goauthentik.io/ldap/group` or in addition to?

BeryJu replied:
I'd say in addition.

Actually, on further inspection, wouldn't it be possible to use `pi.GroupEntry()` and pass and a custom Group object?

Rynoxx replied:
Not a bad idea, I'll do that

Rynoxx replied:
I've ran makemigrations, lint-fix and added the fields to the LDAPProviderForm now

BeryJu commented on outpost/pkg/ldap/utils.go:
```diff
@@ -84,3 +89,31 @@ func (pi *ProviderInstance) GetUserDN(user string) string {
+	}
+
+	dataSize := len(data);
+
```
No quite sure how the logic works, wouldn't it be easier to convert the UUID to its integer form and use the last few digits?

Rynoxx replied:
Yep, it was a bit overcomplicated. I had grand plans for generating the gidNumber at first but scrapped 90% of it, this roundabout way is the scraps from that.

Doing it a simpler way now via hex to bigint conversion to just get the last 5 digits

BeryJu commented on outpost/cmd/ldap/server.go:
```diff
@@ -19,7 +19,12 @@ const helpMessage = `authentik ldap
-- AUTHENTIK_INSECURE: Skip SSL Certificate verification`
+- AUTHENTIK_INSECURE: Skip SSL Certificate verification
+
+Additional environment variables:
```
This should probably be removed

BeryJu reviewed (commented):
LGTM so far, for the newly added fields you need to run `./manage.py makemigrations` to create the database migrations, and then run `make lint-fix` to ensure the formatting is right.

The other thing that could still be done is adding the new fields to the Admin UI in https://github.com/goauthentik/authentik/blob/master/web/src/pages/providers/ldap/LDAPProviderForm.ts, but I can also do that afterwards.

Rynoxx committed:
Refactor code, update documentation and remove obsolete comment

Simplify `GetRIDForGroup`, was a bit overcomplicated before.

Add an additional class/struct `LDAPGroup` which is the new argument
for `pi.GroupEntry` and util functions to create `LDAPGroup` from api.Group and api.User

Add proper support in the interface for changing gidNumber and uidNumber starting points

Rynoxx committed:
make lint-fix for the migration files
